### PR TITLE
Auto generate the config docs page

### LINKED ISSUE #194: Auto generate the config docs page from the code to always keep it in sync
At the moment when we change something in the cli or the cli configs need to remember to also update the page below so quickly becomes out of date.
https://github.com/tellor-io/telliot/blob/master/docs/configuration.md

There is a way to auto generate all cli flags, all possible configs and the current default config.
I think Thanos would be a good start to learn how to do that.
https://github.com/thanos-io/thanos/blob/a11b790ee9984a03bec6031718f962effc7cf45a/Makefile#L155

So the tasks are: auto generate the `configuration` page using a `make docs`
 - display all available cli flags for all commands with an explanation of what each one does
 - display all possible configs  with an explanation of what each field is for
 - display the current default config
 - display the example .env file  with an explanation of what each field is for

Note: 
the cli config page uses gitbook format and here is how it looks like:
https://docs.tellor.io/tellor/telliot/configuration

### PR DESCRIPTION
**Pull request by hhio618**
### Todo
- [x] Generate cli docs from the cli struct.
- [x] Generate config docs from the config struct.
- [x] Add comments to `.env.example`
- [x] Add `help` tags to config structs.
### Refs
closes #194

### CODE REVIEW

File: scripts/cfgdocgen/configuration.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+Telliot commands and config file options are as the following:
+
+#### Required Flags <a id="docs-internal-guid-d1a57725-7fff-a753-9236-759dd3f42eed"></a>
+
+* `--config` \(path to your config file.\)
+
+#### Telliot Commands
+* `--logConfig` \(location of logging config file; default path is current directory\)
+{{range .CliDocs}}
+* `{{ .Name }}` {{range .Arguments }}\({{.Name}}{{if .Optional}}\(optional\){{end}}{{if .Help}}: {{ .Help }}{{end}}\){{end}}  \({{ .Help }}\)
```
Comment (krasi-georgiev) [inline scripts/cfgdocgen/configuration.md:18]: Can you explain how all this works and maybe add comments for someone else looking at this later.

**Review by krasi-georgiev — COMMENTED** [review]
can you also add it as a check to gh actions so that if we change anything in the cli the PR fails unless we also run the script locally to update the docs.

File: scripts/cfgdocgen/configuration.md (hunk 1 above)
Comment (hhio618) [reply scripts/cfgdocgen/configuration.md:18]: Sure I will add some comments.

File: Makefile [hunk 2]
```diff
@@ -80,6 +81,15 @@ build:
 	@[ "${GIT_HASH}" ] || ( echo ">> GIT_HASH is not set"; exit 1 )
 	go build -ldflags "-X main.GitTag=$(GIT_TAG) -X main.GitHash=$(GIT_HASH) -s -w" ./cmd/telliot
 
+.PHONY: generate-config-docs
+generate-config-docs:
```
Comment (krasi-georgiev) [inline Makefile:85]: Lets add a help comment here like for the other commands to explain what this script does exactly.
```suggestion
generate-config-docs: ## Use the example env file and the actual golang config struct to generate the setup doc page. .... <- add more info how it works.
```

File: docs/configuration copy.md [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+## CLI reference
+
+Telliot commands and config file options are as the following:
+
+#### Required Flags <a id="docs-internal-guid-d1a57725-7fff-a753-9236-759dd3f42eed"></a>
+
+* `--config` \(path to your config file.\)
+
+#### Telliot Commands
+* `--logConfig` \(location of logging config file; default path is current directory\)
```
Comment (krasi-georgiev) [inline docs/configuration copy.md:16]: Is this file generated?

Why is this here we don't use `logConfig` anymore?

File: configs/.env.example [hunk 4]
```diff
@@ -1,4 +1,3 @@
-ETH_PRIVATE_KEYS="eeeee6653cdcacc36e3c400ceeeef2aefd59e2642c2f7f298047eeeeeeeeeeee,9643c732204f2a7c9bdb74e2fa08e36d6a4ae8378b983064848b76318fb6507d"
-VIXEOD_KEY=xxx
-CMC_KEY=xxxxxxxxxxxxx
-NODE_WEBSOCKET_URL="wss://mainnet.infura.io/v3/ws/xxxxxxxxxxxxx"
\ No newline at end of file
+ETH_PRIVATE_KEYS="eeeee6653cdcacc36e3c400ceeeef2aefd59e2642c2f7f298047eeeeeeeeeeee,9643c732204f2a7c9bdb74e2fa08e36d6a4ae8378b983064848b76318fb6507d" # required list of private keys separated by `,`   
+$PSR$_KEY=xxxxxxxxxxxxx # API key for getting a specific indexes.json api \(required if you use authenticated API's\)
```
Comment (krasi-georgiev) [inline configs/.env.example:2]: This is not very clear. I think instead we can just add somewhere in the docs that in the API file can use any var from the env file which allows to have auth keys in the API urls.
```suggestion
```

**Review by krasi-georgiev — COMMENTED** [review]
I made huge refactoring of the configs in https://github.com/tellor-io/telliot/pull/440 so would you mind to use that default config struct to generate the doc page and see how it looks like.

File: docs/configuration copy.md (hunk 3 above)
Comment (hhio618) [reply docs/configuration copy.md:16]: Oh, thanks! this file should be removed :+1:

**Comment by hhio618** [comment]
I merged #440 into my PR, now we could test the new config file.

File: Makefile [hunk 5]
```diff
@@ -119,7 +129,7 @@ lint: go-lint shell-lint
 #      --mem-profile-path string   Path to memory profile output file
 # to debug big allocations during linting.
 .PHONY: go-lint
-go-lint: check-git deps $(GOLANGCI_LINT) $(FAILLINT) $(MISSPELL)
+go-lint: check-git deps check-config-docs $(GOLANGCI_LINT) $(FAILLINT) $(MISSPELL)
```
Comment (krasi-georgiev) [inline Makefile:132]: We should not be mixing targets - checking the docs is not linting. Please add it as a separate CI job

File: configs/.env.example (hunk 4 above)
Comment (krasi-georgiev) [reply configs/.env.example:2]: I don't see any change why marked as resolved?

**Review by krasi-georgiev — COMMENTED** [review]
The config file doesn't look different, did you run the auto generate?

File: Makefile (hunk 5 above)
Comment (hhio618) [reply Makefile:132]: Ok!

File: Makefile [hunk 6]
```diff
@@ -82,7 +82,7 @@ build:
 	go build -ldflags "-X main.GitTag=$(GIT_TAG) -X main.GitHash=$(GIT_HASH) -s -w" ./cmd/telliot
 
 .PHONY: generate-config-docs
-generate-config-docs:
+generate-config-docs: ## Use the example env file and the actual golang config struct to generate the setup doc page. .... <- add more info how it works.
```
Comment (krasi-georgiev) [inline Makefile:85]: I meant this only as an example :)  Please make it a proper full comment.

Comment (hhio618) [reply Makefile:85]: Done!

File: Makefile [hunk 7]
```diff
@@ -80,6 +81,15 @@ build:
...
 	go build -ldflags "-X main.GitTag=$(GIT_TAG) -X main.GitHash=$(GIT_HASH) -s -w" ./cmd/telliot
 
+.PHONY: generate-config-docs
+generate-config-docs: ## Use the example env file and the actual golang config struct to generate the setup doc page. .... <- add more info how it works.
+	@go run ./scripts/cfgdocgen --output docs/configuration.md
+
+.PHONY: check-config-docs
+check-config-docs:
+	@go run ./scripts/cfgdocgen --output tmp/configuration.md
+	"$(CMP)" --silent docs/configuration.md tmp/configuration.md || (echo >&2 "outdated docs/configuration.md, run the config docs generator script" ; exit 1)
```
Comment (krasi-georgiev) [inline Makefile:91]: why not use what linting does just by using git, this way will reduce the dependancies.

File: Makefile [hunk 8]
```diff
@@ -13,6 +13,7 @@ GOPROXY     ?= https://proxy.golang.org
 # systems gsed won't be installed, so will use sed as expected.
 SED     ?= $(shell which gsed 2>/dev/null || which sed)
 GIT     ?= $(shell which git)
+CMP     ?= $(shell which cmp)
```
Comment (krasi-georgiev) [inline Makefile:16]: why not use what linting does just by using git, this way will reduce the dependancies.

File: Makefile [hunk 9]
```diff
@@ -80,6 +81,32 @@ build:
...
+generate-config-docs: ## GENERATING CLI DOCS
+generate-config-docs: ## Similar to config struct docs, this will use field annotations from the cli struct and will result in docs entries
+generate-config-docs: ## per cli command consist of a command name (separated by spaces if it's a subcommand), a list of arguments and
+generate-config-docs: ## and an help text from the cli struct `help` annotation.
+generate-config-docs: ## For generating a list of arguments we take a look at the cli struct and find out the fields that contain
+generate-config-docs: ## an arg annotation. we also consider other argument annotations like `optional` and `required` if any. 
+generate-config-docs:	@go run ./scripts/cfgdocgen --output docs/configuration.md
+
+.PHONY: check-config-docs
+check-config-docs:
```
Comment (krasi-georgiev) [inline Makefile:106]: add a comment that is used in the CI to ensure the docs are always up to date.

File: scripts/cfgdocgen/main.go [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+	Required bool
+}
+
+type envDoc struct {
+	Name     string
+	Help     string
+	Required bool
+}
+
+var logger log.Logger
```
Comment (krasi-georgiev) [inline scripts/cfgdocgen/main.go:50]: Lets not use global variables ever! ;)

File: Makefile (hunk 8 above)
Comment (hhio618) [reply Makefile:16]: I didn't saw it, I'll use git instead!

File: scripts/cfgdocgen/main.go (hunk 10 above)
Comment (hhio618) [reply scripts/cfgdocgen/main.go]: Sure :)

File: configs/.env.example [hunk 11]
```diff
@@ -1,4 +1,2 @@
-ETH_PRIVATE_KEYS="eeeee6653cdcacc36e3c400ceeeef2aefd59e2642c2f7f298047eeeeeeeeeeee,9643c732204f2a7c9bdb74e2fa08e36d6a4ae8378b983064848b76318fb6507d"
-VIXEOD_KEY=xxx
-CMC_KEY=xxxxxxxxxxxxx
```
Comment (krasi-georgiev) [inline configs/.env.example:3]: Where is the comment to explain that can use env vars in the API file?

**Review by krasi-georgiev — COMMENTED** [review]
The generated file is missing

File: configs/.env.example (hunk 11 above)
Comment (hhio618) [reply configs/.env.example:3]: Should I put the comment in the generated doc manually, or somewhere else?

File: docs/setup-and-usage.md [hunk 12]
```diff
@@ -127,7 +127,7 @@ wget https://raw.githubusercontent.com/tellor-io/telliot/master/configs/config.j
 ### Create .env file
 
 Most commands require some secrets and these are kept in a separate `configs/.env`. This is a precaution so that are not accidentally exposed as part of the main config. Make a copy of the `env.example` and edit with your secrets.
-For a working setup it is required to at least add one private key in your `"ETH_PRIVATE_KEYS"` environment variable. all of public addresses can be determined from your private keys.
+For a working setup it is required to at least add one private key in your `"ETH_PRIVATE_KEYS"` environment variable. all of public addresses can be determined from your private keys. Also you may need variables in the format of `$PSR$_KEY` for getting a specific indexes.json api \(required if you use authenticated API's\).
```
Comment (krasi-georgiev) [inline docs/setup-and-usage.md:130]: I think the text should be more like: `Any env variable from the .env file is substituted in the API URL. For example......`

File: docs/configuration.md [hunk 13]
```diff
@@ -14,44 +14,118 @@ Telliot commands and config file options are as the following:
...
-* `stake deposit` \(indicates to deposit tokens in the contract\)
-* `stake request` \(indicates you wish to withdraw your stake\)
-* `stake withdraw` \(withdraws your stake, run 1 week after request\)
-* `stake status` \(shows your staking balance\)
-* `balance` \(shows your balance\)
+
+* `accounts`   \(Show accounts\)
+
+
+* `approve` \(Address\)\(Amount\)\(Account\(optional\)\)  \(Approve tokens\)
```
Comment (krasi-georgiev) [inline docs/configuration.md:21]: Hm, I don't think I like this format. Is there no standard formating for cli flags/options we can use here? Not sure how Thanos or the other open source project display the cli args.

File: docs/setup-and-usage.md [hunk 14]
```diff
@@ -128,6 +128,7 @@ wget https://raw.githubusercontent.com/tellor-io/telliot/master/configs/config.j
 
 Most commands require some secrets and these are kept in a separate `configs/.env`. This is a precaution so that are not accidentally exposed as part of the main config. Make a copy of the `env.example` and edit with your secrets.
 For a working setup it is required to at least add one private key in your `"ETH_PRIVATE_KEYS"` environment variable. all of public addresses can be determined from your private keys.
+Any env variable from the .env file is substituted in the API URL (see [indexes.json](https://raw.githubusercontent.com/tellor-io/telliot/master/configs/indexes.json)). for example you may need a `CMC_KEY` env variable for making authenticated CoinMarketCap api calls \(required if you use authenticated API's\).
```
Comment (krasi-georgiev) [inline docs/setup-and-usage.md:131]: actually you can remove it as I will add this in my PR
```suggestion
```

File: cmd/telliot/cli/cmds.go [hunk 15]
```diff
@@ -1,7 +1,7 @@
 // Copyright (c) The Tellor Authors.
 // Licensed under the MIT License.
 
-package main
+package cli
```
Comment (krasi-georgiev) [inline cmd/telliot/cli/cmds.go:4]: I think best to put it in the pgk/cli folder

File: docs/configuration.md (hunk 13 above)
Comment (hhio618) [reply docs/configuration.md:21]: Looks like Thanos uses [embedmd](https://github.com/campoy/embedmd) to embed cli outputs in the docs wherever needed, like this one: https://github.com/thanos-io/thanos/blob/main/docs/components/store.md#flags

Comment (krasi-georgiev) [reply docs/configuration.md:21]: yes that looks ok, lets use that format.

File: cmd/telliot/main.go [hunk 16]
```diff
@@ -4,15 +4,12 @@
...
 import (
-	"fmt"
-
 	"github.com/alecthomas/kong"
 	"github.com/tellor-io/telliot/pkg/cli"
 )
 
 func main() {
 	//lint:ignore faillint it should print to console
-	fmt.Printf(cli.VersionMessage, cli.GitTag, cli.GitHash)
```
Comment (krasi-georgiev) [inline cmd/telliot/main.go]: We still want to show this, maybe as a log, but add it somewhere that will not be printed in the help doc.

File: log.txt [hunk 17]
```diff
@@ -0,0 +1,825 @@
+2021/05/27 00:39:00 ############################################
```
Comment (krasi-georgiev) [inline log.txt:1]: added by mistake

File: docs/configuration.md [hunk 18]
```diff
@@ -34,11 +29,6 @@ Flags:
 
 * `approve`
 ```go
-
-    The official Tellor cli tool  ()
-    -----------------------------------------
-	Website: https://tellor.io
-	Github:  https://github.com/tellor-io/telliot
 Usage: Telliot approve <address> <amount> [<account>]
```
Comment (krasi-georgiev) [inline docs/configuration.md:32]: can we have it lower case `telliot`

Comment (hhio618) [reply docs/configuration.md:32]: Yeah, sure!

File: log.txt (hunk 17 above)
Comment (hhio618) [reply log.txt:1]: Thanks, for letting me know :D

**Comment by krasi-georgiev** [comment]
I just remembered that we also want to print in the docs  what is the default configs.

Should print it in json format so that people can directly copy it to use and just change some things in it.

File: docs/configuration.md [hunk 19]
```diff
@@ -14,45 +14,364 @@ Telliot commands and config file options are as the following:
...
+		"rpc": "info",
+		"tracker": "info"
+	},
+	"ManualDataFile": "configs/manualData.json",
+	"ServerWhitelist": null
+}
+```
 ### LogConfig file options
 
 The logging.config file consists of two fields: \* component \* level
```
Comment (krasi-georgiev) [inline docs/configuration.md:377]: We changed that and the config for each component has a field for the log level.

Comment (hhio618) [reply docs/configuration.md]: Done!

**Comment by krasi-georgiev** [comment]
Ok I think this looks good, what is that last item in your list? - Add help tags to config structs.

**Comment by hhio618** [comment]
Thanks, the last one was about adding help annotation the config struct fields.

**Review by krasi-georgiev — APPROVED** [review]
